servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -43,7 +52,66 @@ private DefaultAsyncContextProvider() {
     @Nonnull
     @Override
     public ContextMap context() {
-        return CONTEXT_LOCAL.get();
+        final Thread t = Thread.currentThread();
+        if (t instanceof ContextMapHolder) {
+            final ContextMapHolder contextMapHolder = (ContextMapHolder) t;
+            ContextMap map = contextMapHolder.context();
+            if (map == null) {
+                map = newContextMap();
+                contextMapHolder.context(map);
+            }
+            return map;
+        } else {
+            return CONTEXT_THREAD_LOCAL.get();
+        }
+    }
+
+    @Override
+    public ContextMap saveContext() {
+        return context();
+    }
+
+    @Override
+    public Scope attachContext(ContextMap contextMap) {
+        final Thread currentThread = Thread.currentThread();
+        if (currentThread instanceof ContextMapHolder) {
+            final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
+            ContextMap prev = asyncContextMapHolder.context();
+            asyncContextMapHolder.context(contextMap);
+            return prev == null ? Scope.NOOP : () -> detachContext(contextMap, prev);
```

Comment:
Even if the `prev == null`, should we still restore the previous state of the `ContextMapHolder` to be `null`? Is there any risk of unexpected state propagation if we don't clean after the task is done?

Reply:
Let's always restore it. It's better safe than sorry.